olive/passes/pytorch/gptq.py

```diff
@@ -0,0 +1,161 @@
+            ),
+            "true_sequential": PassConfigParam(
+                type_=bool,
+                default_value=False,
+                description="Use true sequential for quantization. Default value is False.",
+            ),
+            "desc_act": PassConfigParam(
+                type_=bool,
+                default_value=False,
+                description="Use descriptive activation for quantization. Default value is False.",
+            ),
+            "sym": PassConfigParam(
+                type_=bool,
+                default_value=False,
+                description="Symmetric quantization. Default value is True.",
+            ),
+            "gpu": PassConfigParam(
+                type_=bool,
+                default_value=True,
+                description="Use GPU for quantization. Default value is True.",
+            ),
+            "export_fp16": PassConfigParam(
+                type_=bool,
+                default_value=True,
+                description="Export FP16 onnx model. Default value is True.",
+            ),
+            "export_optimization": PassConfigParam(
+                type_=str,
+                default_value=None,
+                description="Export optimization level. Default value is None.",
+            ),
+            "enable_transformers_specific_optimizations": PassConfigParam(
+                type_=bool,
+                default_value=False,
+                description="Enable transformers specific optimizations. Default value is False.",
+            ),
+        }
+        config.update(get_external_data_config())
+        return config
+
+    @torch.no_grad()
+    def _run_for_config(
+        self, model: PyTorchModelHandler, data_root: str, config: Dict[str, Any], output_model_path: str
+    ) -> PyTorchModelHandler:
+        from optimum.exporters.onnx import onnx_export_from_model  # type: ignore[import]
+        from transformers import AutoModelForCausalLM, AutoTokenizer, GPTQConfig  # type: ignore[import]
+
+        tokenizer = AutoTokenizer.from_pretrained(model.hf_config.model_name, use_fast=False)
+
+        gptq_config = GPTQConfig(
+            bits=config["bits"],
+            dataset=config["dataset"],
+            group_size=config["group_size"],
+            damp_percent=config["damp_percent"],
+            static_groups=config["static_groups"],
+            true_sequential=config["true_sequential"],
+            desc_act=config["desc_act"],
+            sym=config["sym"],
+            tokenizer=tokenizer,
+        )
+
+        from unittest import mock
+
+        def get_onnx_quant_linear(*args, **kwargs):
+            return QuantLinearORT
+
+        # GPTQ Quantization in transformers use auto_gptq under the hood, so we
+        # replace QuantLinear in auto_gptq with QuantLinearORT for quant linear layer packing
+        with mock.patch("auto_gptq.utils.import_utils.dynamically_import_QuantLinear", get_onnx_quant_linear):
```

Comment:
why the production code has UT code?

Reply:
It is used to dynamic repalce 3rdlib func, I tried other method like subclass but they are not easy to implemented. This mock metho only replace the 3rdlib func in the with block, it will not influence other code.

Reply:
It is not a good practice to use this way to replace the underlying function.

Reply:
could the following line works?
optimum.gptq.quantizer.dynamically_import_QuantLinear = get_onnx_quant_linear

Reply:
Tried several weeks ago, I'll try this again and maybe other ways to fix here.